Extension/src/LanguageServer/Providers/findAllReferencesProvider.ts

```diff
@@ -3,98 +3,63 @@
-            };
+    public async provideReferences(document: vscode.TextDocument, position: vscode.Position, context: vscode.ReferenceContext, token: vscode.CancellationToken):
+        Promise<vscode.Location[] | undefined> {
+        await this.client.awaitUntilLanguageClientReady();
+        workspaceReferences.cancelCurrentReferenceRequest(CancellationSender.NewRequest);
+
+        // Listen to a cancellation for this request. When this request is cancelled,
+        // use a local cancellation source to explicitly cancel a token.
+        const cancelSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+        const cancellationTokenListener: vscode.Disposable = token.onCancellationRequested(() => { cancelSource.cancel(); });
+        const requestCanceledListener: vscode.Disposable = workspaceReferences.onCancellationRequested(sender => { cancelSource.cancel(); });
 
-            if (DefaultClient.referencesRequestPending || (workspaceReferences.symbolSearchInProgress && !workspaceReferences.referencesRefreshPending)) {
-                const cancelling: boolean = DefaultClient.referencesPendingCancellations.length > 0;
-                DefaultClient.referencesPendingCancellations.push({
-                    reject: () => {
-                        // Complete with nothing instead of rejecting, to avoid an error message from VS Code
-                        const locations: vscode.Location[] = [];
-                        resolve(locations);
-                    }, callback
-                });
-                if (!cancelling) {
-                    DefaultClient.renamePending = false;
-                    workspaceReferences.referencesCanceled = true;
-                    if (!DefaultClient.referencesRequestPending) {
-                        workspaceReferences.referencesCanceledWhilePreviewing = true;
-                    }
-                    this.client.languageClient.sendNotification(CancelReferencesNotification);
+        // Send the request to the language server.
+        const locationsResult: vscode.Location[] = [];
+        const params: ReferencesParams = {
+            newName: "",
+            position: Position.create(position.line, position.character),
+            textDocument: { uri: document.uri.toString() }
+        };
+        const response: ReferencesResult = await this.client.languageClient.sendRequest(FindAllReferencesRequest, params, cancelSource.token);
+
+        // Reset anything that can be cleared before procossing the result.
+        // Note: ReferencesManager.resetReferences is called in ReferencesManager.showResultsInPanelView
+        workspaceReferences.resetProgressBar();
+        cancellationTokenListener.dispose();
+        requestCanceledListener.dispose();
+
+        // Process the result.
+        if (cancelSource.token.isCancellationRequested || response.referenceInfos === null || response.isCanceled) {
+            throw new vscode.CancellationError();
```

Comment:
This appears to cause VS Code to show `Exception has occurred: TypeError: Cannot destructure property 'range' of 'e.location' as it is undefined.`. Can you change it to `return undefined;`? That seemed to fix it for me. I'm not sure where it's documented about the usage of vscode.CancellationError or if we should report a VS Code bug on it or not.

The error is from code at
https://github.com/microsoft/vscode/blob/69246279e5b60be0f3cfd77da6f64a211a55e64a/extensions/references-view/src/references/model.ts#L288

Reply:
This is a regression compared to 1.15.4 -- I think we should return undefined for all the cases that VS Code shows an error popup

Reply:
Added comment to clarify usage of undefined.